{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "04-Jul-2019 17:40:57 INFO GEOparse - Parsing ./GSE107011_family.soft.gz: \n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SERIES: GSE107011\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - PLATFORM: GPL11154\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859411\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859412\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859413\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859414\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859415\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859416\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859417\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859418\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859419\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859420\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859421\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859422\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859423\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859424\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859425\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859426\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859427\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859428\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859429\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859430\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859431\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859432\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859433\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859434\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859435\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859436\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859437\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859438\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859439\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859440\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859441\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859442\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859443\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859444\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859445\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859446\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859447\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859448\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859449\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859450\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859451\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859452\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859453\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859454\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859455\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859456\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859457\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859458\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859459\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859460\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859461\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859462\n",
      "04-Jul-2019 17:40:57 DEBUG GEOparse - SAMPLE: GSM2859463\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859464\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859465\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859466\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859467\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859468\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859469\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859470\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859471\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859472\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859473\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859474\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859475\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859476\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859477\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859478\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859479\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859480\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859481\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859482\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859483\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859484\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859485\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859486\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859487\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859488\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859489\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859490\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859491\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859492\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859493\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859494\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859495\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859496\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859497\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859498\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859499\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859500\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859501\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859502\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859503\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859504\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859505\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859506\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859507\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859508\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859509\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859510\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859511\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859512\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859513\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859514\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859515\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859516\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859517\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859518\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859519\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859520\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859521\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859522\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859523\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859524\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859525\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859526\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859527\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859528\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859529\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859530\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859531\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859532\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859533\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859534\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859535\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859536\n",
      "04-Jul-2019 17:40:58 DEBUG GEOparse - SAMPLE: GSM2859537\n"
     ]
    }
   ],
   "source": [
    "gse = GEOparse.get_GEO(filepath=\"./GSE107011_family.soft.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GSM example:\n",
      "Name:  GSM2859411\n",
      "Metadata:\n",
      " - title : DZQV_CD8_naive_rep4\n",
      " - geo_accession : GSM2859411\n",
      " - status : Public on Feb 06 2019\n",
      " - submission_date : Nov 16 2017\n",
      " - last_update_date : May 15 2019\n",
      " - type : SRA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : Naive CD8 T cells\n",
      " - organism_ch1 : Homo sapiens\n",
      " - taxid_ch1 : 9606\n",
      " - characteristics_ch1 : cell type: Naive CD8 T cells, disease status: healthy, gender: M\n",
      " - molecule_ch1 : polyA RNA\n",
      " - extract_protocol_ch1 : RNA was extracted with TRIzol reagent, followed by clean-up and DNase I treatment with QIAGEN RNeasy mini kit in accordance with the prescribed protocol provided with the kit. Quality control was performed with Agilent Bioanalyser. The RNA concentration was determined using the Quant-iTTM RiboGreen® RNA Assay Kit, The cDNA libraries were prepared from 2 ng of total RNA and 1 μl of a 1:50,000 dilution of external RNA control consortium (ERCC) spike-in control mix (Thermo Fisher Scientific) using SMARTSeq v2 protocol (Picelli et al., 2014) with the following modifications: 1) use of 20µM template-switching oligos (TSO), 2) use of 250 pg of cDNA with 1/5 reaction mixtures of Illumina Nextera XT kit.\n",
      " - description : cell_number (facs): 900000, total_rna_yield (ng): 321.78, DZQV_CD8_naive\n",
      " - data_processing : Sequenced reads were trimmed for adaptor sequence, and masked for low-complexity or low-quality sequence, then mapped to the transcriptome with kallisto, Transcirpt per million (TPM) values of all trancripts were obtained with kallisto. Expression counts at gene level were obtained by summarizing transcript expression values., Genome_build: GENCODE 26, Supplementary_files_format_and_content: tab-delimited text files include TPM values for each sample\n",
      " - platform_id : GPL11154\n",
      " - contact_name : Gianni,,Monaco\n",
      " - contact_email : mongianni1@gmail.com\n",
      " - contact_institute : University of Liverpool\n",
      " - contact_address : 6 West Derby Street\n",
      " - contact_city : Liverpool\n",
      " - contact_zip/postal_code : L78TX\n",
      " - contact_country : United Kingdom\n",
      " - instrument_model : Illumina HiSeq 2000\n",
      " - library_selection : cDNA\n",
      " - library_source : transcriptomic\n",
      " - library_strategy : RNA-Seq\n",
      " - relation : BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN08035215, SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX3399186\n",
      " - supplementary_file_1 : NONE\n",
      " - series_id : GSE107011, GSE107019\n",
      " - data_row_count : 0\n",
      "Table data:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "print()\n",
    "print(\"GSM example:\")\n",
    "for gsm_name, gsm in gse.gsms.items():\n",
    "    print(\"Name: \", gsm_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gsm.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    print (\"Table data:\",)\n",
    "    print (gsm.table.head())\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GPL example:\n",
      "Name:  GPL11154\n",
      "Metadata:\n",
      " - title : Illumina HiSeq 2000 (Homo sapiens)\n",
      " - geo_accession : GPL11154\n",
      " - status : Public on Nov 02 2010\n",
      " - submission_date : Nov 02 2010\n",
      " - last_update_date : Mar 27 2019\n",
      " - technology : high-throughput sequencing\n",
      " - distribution : virtual\n",
      " - organism : Homo sapiens\n",
      " - taxid : 9606\n",
      " - contact_name : ,,GEO\n",
      " - contact_country : USA\n",
      " - data_row_count : 0\n",
      "Table data:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "print()\n",
    "print(\"GPL example:\")\n",
    "for gpl_name, gpl in gse.gpls.items():\n",
    "    print(\"Name: \", gpl_name) \n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gpl.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value))) \n",
    "    print(\"Table data:\",)\n",
    "    print(gpl.table.head())\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<SAMPLE: GSM2859411>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gse.gsms['GSM2859411']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on GSM in module GEOparse.GEOTypes object:\n",
      "\n",
      "class GSM(SimpleGEO)\n",
      " |  Class that represents sample from GEO database.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      GSM\n",
      " |      SimpleGEO\n",
      " |      BaseGEO\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  annotate(self, gpl, annotation_column, gpl_on='ID', gsm_on='ID_REF', in_place=False)\n",
      " |      Annotate GSM with provided GPL\n",
      " |      \n",
      " |      Args:\n",
      " |          gpl (:obj:`pandas.DataFrame`): A Platform or DataFrame to annotate with\n",
      " |          annotation_column (str`): Column in a table for annotation\n",
      " |          gpl_on (:obj:`str`): Use this column in GSM to merge. Defaults to \"ID\".\n",
      " |          gsm_on (:obj:`str`): Use this column in GPL to merge.\n",
      " |              Defaults to \"ID_REF\".\n",
      " |          in_place (:obj:`bool`): Substitute table in GSM by new annotated\n",
      " |              table. Defaults to False.\n",
      " |      \n",
      " |      Returns:\n",
      " |          :obj:`pandas.DataFrame` or :obj:`None`: Annotated table or None\n",
      " |      \n",
      " |      \n",
      " |      Raises:\n",
      " |          TypeError: GPL should be GPL or pandas.DataFrame\n",
      " |  \n",
      " |  annotate_and_average(self, gpl, expression_column, group_by_column, rename=True, force=False, merge_on_column=None, gsm_on=None, gpl_on=None)\n",
      " |      Annotate GSM table with provided GPL.\n",
      " |      \n",
      " |      Args:\n",
      " |          gpl (:obj:`GEOTypes.GPL`): Platform for annotations\n",
      " |          expression_column (:obj:`str`): Column name which \"expressions\"\n",
      " |              are represented\n",
      " |          group_by_column (:obj:`str`): The data will be grouped and averaged\n",
      " |              over this column and only this column will be kept\n",
      " |          rename (:obj:`bool`): Rename output column to the\n",
      " |              self.name. Defaults to True.\n",
      " |          force (:obj:`bool`): If the name of the GPL does not match the platform\n",
      " |              name in GSM proceed anyway. Defaults to False.\n",
      " |          merge_on_column (:obj:`str`): Column to merge the data\n",
      " |              on. Defaults to None.\n",
      " |          gsm_on (:obj:`str`): In the case columns to merge are different in GSM\n",
      " |              and GPL use this column in GSM. Defaults to None.\n",
      " |          gpl_on (:obj:`str`): In the case columns to merge are different in GSM\n",
      " |              and GPL use this column in GPL. Defaults to None.\n",
      " |      \n",
      " |      Returns:\n",
      " |          :obj:`pandas.DataFrame`: Annotated data\n",
      " |  \n",
      " |  download_SRA(self, email, directory='./', **kwargs)\n",
      " |      Download RAW data as SRA file.\n",
      " |      \n",
      " |      The files will be downloaded to the sample directory created ad hoc\n",
      " |      or the directory specified by the parameter. The sample has to come\n",
      " |      from sequencing eg. mRNA-seq, CLIP etc.\n",
      " |      \n",
      " |      An important parameter is a filetype. By default an SRA\n",
      " |      is accessed by FTP and such file is downloaded. This does not\n",
      " |      require additional libraries. However in order\n",
      " |      to produce FASTA of FASTQ files one would need to use SRA-Toolkit.\n",
      " |      Thus, it is assumed that this library is already installed or it\n",
      " |      will be installed in the near future. One can immediately specify\n",
      " |      the download type to fasta or fastq.\n",
      " |      \n",
      " |      To see all possible ``**kwargs`` that could be passed to the function\n",
      " |      see the description of :class:`~GEOparse.sra_downloader.SRADownloader`.\n",
      " |      \n",
      " |      Args:\n",
      " |          email (:obj:`str`): an email (any) - Required by NCBI for access\n",
      " |          directory (:obj:`str`, optional): The directory to which download\n",
      " |              the data. Defaults to \"./\".\n",
      " |          **kwargs: Arbitrary keyword arguments, see description\n",
      " |      \n",
      " |      Returns:\n",
      " |          :obj:`dict`: A dictionary containing only one key (``SRA``) with\n",
      " |              the list of downloaded files.\n",
      " |      \n",
      " |      Raises:\n",
      " |          :obj:`TypeError`: Type to download unknown\n",
      " |          :obj:`NoSRARelationException`: No SRAToolkit\n",
      " |          :obj:`Exception`: Wrong e-mail\n",
      " |          :obj:`HTTPError`: Cannot access or connect to DB\n",
      " |  \n",
      " |  download_supplementary_files(self, directory='./', download_sra=True, email=None, sra_kwargs=None)\n",
      " |      Download all supplementary data available for the sample.\n",
      " |      \n",
      " |      Args:\n",
      " |          directory (:obj:`str`): Directory to download the data (in this directory\n",
      " |              function will create new directory with the files).\n",
      " |              Defaults to \"./\".\n",
      " |          download_sra (:obj:`bool`): Indicates whether to download SRA raw\n",
      " |              data too. Defaults to True.\n",
      " |          email (:obj:`str`): E-mail that will be provided to the Entrez.\n",
      " |              It is mandatory if download_sra=True. Defaults to None.\n",
      " |          sra_kwargs (:obj:`dict`, optional): Kwargs passed to the\n",
      " |              download_SRA method. Defaults to None.\n",
      " |      \n",
      " |      Returns:\n",
      " |          :obj:`dict`: A key-value pair of name taken from the metadata and\n",
      " |              paths downloaded, in the case of SRA files the key is ``SRA``.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  geotype = 'SAMPLE'\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from SimpleGEO:\n",
      " |  \n",
      " |  __init__(self, name, metadata, table, columns)\n",
      " |      Initialize simple GEO object.\n",
      " |      \n",
      " |      Args:\n",
      " |          name (:obj:`str`): Name of the object\n",
      " |          metadata (:obj:`dict`): Metadata information\n",
      " |          table (:obj:`pandas.DataFrame`): Table with the data from SOFT file\n",
      " |          columns (:obj:`pandas.DataFrame`): Description of the columns,\n",
      " |              number of columns, order and names represented as index\n",
      " |              in this DataFrame has to be the same as table.columns.\n",
      " |      \n",
      " |      Raises:\n",
      " |          :obj:`ValueError`: Table should be a DataFrame\n",
      " |          :obj:`ValueError`: Columns' description should be a DataFrame\n",
      " |          :obj:`DataIncompatibilityException`: Columns are wrong\n",
      " |          :obj:`ValueError`: Description has to be present in columns\n",
      " |  \n",
      " |  head(self)\n",
      " |      Print short description of the object.\n",
      " |  \n",
      " |  show_columns(self)\n",
      " |      Print columns in SOFT format.\n",
      " |  \n",
      " |  show_table(self, number_of_lines=5)\n",
      " |      Show few lines of the table the table as pandas.DataFrame.\n",
      " |      \n",
      " |      Args:\n",
      " |          number_of_lines (:obj:`int`): Number of lines to show. Defaults to 5.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from SimpleGEO:\n",
      " |  \n",
      " |  __metaclass__ = <class 'abc.ABCMeta'>\n",
      " |      Metaclass for defining Abstract Base Classes (ABCs).\n",
      " |      \n",
      " |      Use this metaclass to create an ABC.  An ABC can be subclassed\n",
      " |      directly, and then acts as a mix-in class.  You can also register\n",
      " |      unrelated concrete classes (even built-in classes) and unrelated\n",
      " |      ABCs as 'virtual subclasses' -- these and their descendants will\n",
      " |      be considered subclasses of the registering ABC by the built-in\n",
      " |      issubclass() function, but the registering ABC won't show up in\n",
      " |      their MRO (Method Resolution Order) nor will method\n",
      " |      implementations defined by the registering ABC be callable (not\n",
      " |      even via super()).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from BaseGEO:\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  get_accession(self)\n",
      " |      Return accession ID of the sample.\n",
      " |      \n",
      " |      Returns:\n",
      " |          :obj:`str`: GEO accession ID\n",
      " |  \n",
      " |  get_metadata_attribute(self, metaname)\n",
      " |      Get the metadata attribute by the name.\n",
      " |      \n",
      " |      Args:\n",
      " |          metaname (:obj:`str`): Name of the attribute\n",
      " |      \n",
      " |      Returns:\n",
      " |          :obj:`list` or :obj:`str`: Value(s) of the requested metadata\n",
      " |              attribute\n",
      " |      \n",
      " |      Raises:\n",
      " |          NoMetadataException: Attribute error\n",
      " |          TypeError: Metadata should be a list\n",
      " |  \n",
      " |  get_type(self)\n",
      " |      Get the type of the GEO.\n",
      " |      \n",
      " |      Returns:\n",
      " |          :obj:`str`: Type attribute of the GEO\n",
      " |  \n",
      " |  show_metadata(self)\n",
      " |      Print metadata in SOFT format.\n",
      " |  \n",
      " |  to_soft(self, path_or_handle, as_gzip=False)\n",
      " |      Save the object in a SOFT format.\n",
      " |      \n",
      " |      Args:\n",
      " |          path_or_handle (:obj:`str` or :obj:`file`): Path or handle to\n",
      " |              output file\n",
      " |          as_gzip (:obj:`bool`): Save as gzip\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from BaseGEO:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(gse.gsms['GSM2859411'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__class__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__metaclass__',\n",
       " '__module__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_get_columns_as_string',\n",
       " '_get_metadata_as_string',\n",
       " '_get_object_as_soft',\n",
       " '_get_table_as_string',\n",
       " 'annotate',\n",
       " 'annotate_and_average',\n",
       " 'columns',\n",
       " 'download_SRA',\n",
       " 'download_supplementary_files',\n",
       " 'geotype',\n",
       " 'get_accession',\n",
       " 'get_metadata_attribute',\n",
       " 'get_type',\n",
       " 'head',\n",
       " 'metadata',\n",
       " 'name',\n",
       " 'relations',\n",
       " 'show_columns',\n",
       " 'show_metadata',\n",
       " 'show_table',\n",
       " 'table',\n",
       " 'to_soft']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(gse.gsms['GSM2859411'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [description]\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gse.gsms['GSM2859411'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method GSM.download_supplementary_files of <SAMPLE: GSM2859411>>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gse.gsms['GSM2859411'].download_supplementary_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
